core/state/access_witness.go

```diff
@@ -110,11 +110,25 @@ func (aw *AccessWitness) TouchAndChargeValueTransfer(callerAddr, targetAddr []by
 	return gas
 }
 
+// TouchAndChargeContractCreateCheck charges access costs before
+// a contract creation is initiated. It is just reads, because the
+// address collision is done before the transfer, and so no write
+// are guaranteed to happen at this point.
+func (aw *AccessWitness) TouchAndChargeContractCreateCheck(addr []byte) uint64 {
+	var gas uint64
+	gas += aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BasicDataLeafKey, false)
+	gas += aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.CodeHashLeafKey, false)
+	return gas
+}
+
 // TouchAndChargeContractCreateInit charges access costs to initiate
-// a contract creation
-func (aw *AccessWitness) TouchAndChargeContractCreateInit(addr []byte, createSendsValue bool) uint64 {
+// a contract creation. It writes to the balance if a value transfer
+// happens, and to the code hash leaf since it will be created no
+// matter what.
+func (aw *AccessWitness) TouchAndChargeContractCreateInit(addr []byte, sendsValue bool) uint64 {
 	var gas uint64
 	gas += aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.BasicDataLeafKey, true)
+	gas += aw.touchAddressAndChargeGas(addr, zeroTreeIndex, utils.CodeHashLeafKey, sendsValue)
```

Comment:
at this point, we know that the account does not exist, but if the value is non-zero, then it will still be created even if the _contract_ creation fails. So we need to charge the write/fill costs if the value is non-zero.